### PR TITLE
Merged Subgraph

### PR DESCRIPTION
**Pull request by ejwessel**
## Description
https://github.com/airswap/airswap-protocols/issues/505
AirSwap Subgraph to help reduce costs from Alchemy and Infura

Quick description:

- [ ] N/A
- [X] New features
- [ ] Bug fixes
- [ ] Typo/small tweaks

## Changes
Implemented Entities and Mappings for:
- Swap
- Indexer
- Index
- DelegateFactory
- Delegate

## Tests
1. local ethereum node running
```
yarn ganache -h 0.0.0.0
```
2. deploy and grab addresses
3. update the config with addresses
4. local graph node running
```
docker-compose up
```
5. deploy AirSwap subgraph locally: 
```
yarn prepare:development && yarn codegen && yarn remove-local;  yarn create-local && yarn deploy-local
```
6. query

### CODE REVIEW

File: source/delegate/test/Delegate.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 const Delegate = artifacts.require('Delegate')
+const DelegateFactory = artifacts.require('DelegateFactory')
```
Comment (ejwessel) [inline source/delegate/test/Delegate.js:2]: Necessary in order to capture created Delegates and run queries against them.

**Comment by ejwessel** [comment]
Probably need to write the integration tests for Indexer when it creates and Index.

**Comment by ejwessel** [comment]
query
```
{
    users {
        id
        authorizedSigners {
            id
        }
        authorizedSenders {
            id
        }
        executedOrders
        cancelledNonces
    }
    executedOrders {
        id
        signer {
            id
        }
        sender {
            id
        }
        swap {
            id
        }
    }
}
```
results
```
{
    "data": {
        "executedOrders": [
            {
                "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b41586557185708",
                "sender": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "signer": {
                    "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b4"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b41586557185846",
                "sender": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "signer": {
                    "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b4"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557178648",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557180385",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557180937",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557181908",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557182889",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557183633",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557184674",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            },
            {
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa1586557185128",
                "sender": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "signer": {
                    "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
                },
                "swap": {
                    "id": "0x2711824d10c7935d247fe1189b336d4c7e6ed918"
                }
            }
        ],
        "users": [
            {
                "authorizedSenders": [],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0x0000000000000000000000000000000000000000"
            },
            {
                "authorizedSenders": [],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0x75c4fb2e81a6d3420125f5145182f528d1699146"
            },
            {
                "authorizedSenders": [],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b4"
            },
            {
                "authorizedSenders": [
                    {
                        "id": "0xc91579bb7972f76d595f8665bffaf92874c8084c"
                    }
                ],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
            },
            {
                "authorizedSenders": [],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0xc91579bb7972f76d595f8665bffaf92874c8084c"
            },
            {
                "authorizedSenders": [],
                "authorizedSigners": [
                    {
                        "id": "0x9d2fb0bcc90c6f3fa3a98d2c760623a4f6ee59b4"
                    },
                    {
                        "id": "0xc91579bb7972f76d595f8665bffaf92874c8084c"
                    }
                ],
                "cancelledNonces": [
                    "0",
                    "1",
                    "2",
                    "3"
                ],
                "executedOrders": [],
                "id": "0xe092b1fa25df5786d151246e492eed3d15ea4daa"
            },
            {
                "authorizedSenders": [],
                "authorizedSigners": [],
                "cancelledNonces": [],
                "executedOrders": [],
                "id": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf"
            }
        ]
    }
}
```

**Comment by ejwessel** [comment]
query
```
{
    tokens {
        id
        isBlacklisted
    }
    indexers {
        id
    }
    indexContracts {
        id
        indexer {
            id
        }
        signerToken {
            id
        }
        senderToken {
            id
        }
        protocol
    }
    stakedAmounts {
        id
        indexer {
            id
        }
        signerToken {
            id
        }
        senderToken {
            id
        }
        staker {
            id
        }
        protocol
        stakeAmount
    }
    locators {
        id
        owner {
            id
        }
        index {
            id
        }
        score
        locator
    }
}
```
result
```
{
    "data": {
        "indexContracts": [
            {
                "id": "0x23b928496d14b6df28a50287e8267ddb1f4ad68f",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x1234",
                "senderToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                },
                "signerToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                }
            },
            {
                "id": "0x4dfc6ef9989de3d2ad2bb05e2d12268a7aec0ccf",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x9999",
                "senderToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                },
                "signerToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                }
            },
            {
                "id": "0xb12873c58471d610a09f065391898bab7a78478d",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x1234",
                "senderToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                },
                "signerToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                }
            }
        ],
        "indexers": [
            {
                "id": "0x55a67cba9119ad62a773e76d7d3dc5d48e6720f6"
            },
            {
                "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
            }
        ],
        "locators": [
            {
                "id": "0x75c4fb2e81a6d3420125f5145182f528d16991460x23b928496d14b6df28a50287e8267ddb1f4ad68f",
                "index": {
                    "id": "0x23b928496d14b6df28a50287e8267ddb1f4ad68f"
                },
                "locator": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf000000000000000000000000",
                "owner": {
                    "id": "0x75c4fb2e81a6d3420125f5145182f528d1699146"
                },
                "score": "500"
            },
            {
                "id": "0xc0d8f541ab8b71f20c10261818f2f401e81940490xb12873c58471d610a09f065391898bab7a78478d",
                "index": {
                    "id": "0xb12873c58471d610a09f065391898bab7a78478d"
                },
                "locator": "0xb1d1d474bf38ead491a0bc62ec97d96484ef7588000000000000000000000000",
                "owner": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "score": "500"
            },
            {
                "id": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf0xb12873c58471d610a09f065391898bab7a78478d",
                "index": {
                    "id": "0xb12873c58471d610a09f065391898bab7a78478d"
                },
                "locator": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf000000000000000000000000",
                "owner": {
                    "id": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf"
                },
                "score": "50"
            }
        ],
        "stakedAmounts": [
            {
                "id": "0x75c4fb2e81a6d3420125f5145182f528d16991460x7959fb1645f94899d37879c10faa6081803b0762",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x1234",
                "senderToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                },
                "signerToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                },
                "stakeAmount": "500",
                "staker": {
                    "id": "0x75c4fb2e81a6d3420125f5145182f528d1699146"
                }
            },
            {
                "id": "0xc0d8f541ab8b71f20c10261818f2f401e81940490x7959fb1645f94899d37879c10faa6081803b0762",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x1234",
                "senderToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                },
                "signerToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                },
                "stakeAmount": "500",
                "staker": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                }
            },
            {
                "id": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf0x7959fb1645f94899d37879c10faa6081803b0762",
                "indexer": {
                    "id": "0x7959fb1645f94899d37879c10faa6081803b0762"
                },
                "protocol": "0x1234",
                "senderToken": {
                    "id": "0xcb15681705864bff0c108a43862e5de9a414738a"
                },
                "signerToken": {
                    "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812"
                },
                "stakeAmount": "50",
                "staker": {
                    "id": "0xf1f8ef6b4d4ba31079e2263ec85c03fd5a0802bf"
                }
            }
        ],
        "tokens": [
            {
                "id": "0x0000000000000000000000000000000000000000",
                "isBlacklisted": false
            },
            {
                "id": "0x12aca48b130a466e2114ef5a6ef91f3bf3ed89b2",
                "isBlacklisted": false
            },
            {
                "id": "0x29fc864fc75e6dff142dd3da26103027f4ec3794",
                "isBlacklisted": false
            },
            {
                "id": "0xabb93815dfc8c62fe73fffad83ea66480d7a2812",
                "isBlacklisted": false
            },
            {
                "id": "0xcb15681705864bff0c108a43862e5de9a414738a",
                "isBlacklisted": false
            },
            {
                "id": "0xe08c90f6263789fc5bca6195f8f7eddd991a01d5",
                "isBlacklisted": false
            },
            {
                "id": "0xf273803f6b710ba4c3aa1fcf050671ad899b1460",
                "isBlacklisted": false
            }
        ]
    }
}
```

**Comment by ejwessel** [comment]
query
```
{
    delegateFactories {
        id
    }
    delegateContracts {
        id
        factory {
            id
        }
        swap {
            id
        }
        indexer {
            id
        }
        owner {
            id
        }
        tradeWallet
    }
    rules {
        id
        delegate {
            id
        }
        owner {
            id
        }
        signerToken {
            id
        }
        senderToken {
            id
        }
        maxSenderAmount
        priceCoef
        priceExp
    }
}
```
result
```
{
    "data": {
        "delegateContracts": [
            {
                "factory": {
                    "id": "0x8b01ffee1abb64596fede226ce253703220d82c7"
                },
                "id": "0x49259ceb3fd5560518c078c406e802f3b2447eff",
                "indexer": {
                    "id": "0x55a67cba9119ad62a773e76d7d3dc5d48e6720f6"
                },
                "owner": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "swap": {
                    "id": "0xb4a15a1e1ad0a7daef3ffad9d385cc98cbb38539"
                },
                "tradeWallet": "0x6f03947036cba3279b07cd6ea5ca674ca51e52ba"
            }
        ],
        "delegateFactories": [
            {
                "id": "0x8b01ffee1abb64596fede226ce253703220d82c7"
            }
        ],
        "rules": [
            {
                "delegate": {
                    "id": "0x49259ceb3fd5560518c078c406e802f3b2447eff"
                },
                "id": "0x49259ceb3fd5560518c078c406e802f3b2447eff0x29fc864fc75e6dff142dd3da26103027f4ec37940xe08c90f6263789fc5bca6195f8f7eddd991a01d5",
                "maxSenderAmount": "10000",
                "owner": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "priceCoef": "5",
                "priceExp": "3",
                "senderToken": {
                    "id": "0x29fc864fc75e6dff142dd3da26103027f4ec3794"
                },
                "signerToken": {
                    "id": "0xe08c90f6263789fc5bca6195f8f7eddd991a01d5"
                }
            },
            {
                "delegate": {
                    "id": "0x49259ceb3fd5560518c078c406e802f3b2447eff"
                },
                "id": "0x49259ceb3fd5560518c078c406e802f3b2447eff0xe08c90f6263789fc5bca6195f8f7eddd991a01d50x29fc864fc75e6dff142dd3da26103027f4ec3794",
                "maxSenderAmount": "2",
                "owner": {
                    "id": "0xc0d8f541ab8b71f20c10261818f2f401e8194049"
                },
                "priceCoef": "100",
                "priceExp": "0",
                "senderToken": {
                    "id": "0xe08c90f6263789fc5bca6195f8f7eddd991a01d5"
                },
                "signerToken": {
                    "id": "0x29fc864fc75e6dff142dd3da26103027f4ec3794"
                }
            }
        ]
    }
}
```

File: tools/deployer/AirSwap/package.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "AirSwap",
```
Comment (ejwessel) [inline tools/deployer/AirSwap/package.json:2]: @airswap/subgraph

File: tools/deployer/AirSwap/package.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  "version": "0.0.1",
+  "license": "UNLICENSED",
```
Comment (ejwessel) [inline tools/deployer/AirSwap/package.json:4]: Apache-2.0

**Comment by dmosites** [comment]
Let's place this in `tools` / `subgraph` and remove its `yarn.lock` to wire it into the lerna setup.

File: .gitignore [hunk 4]
```diff
@@ -10,6 +10,5 @@ build
 flatten/
 .env
 venv
-build
```
Comment (ejwessel) [inline .gitignore:13]: this was a duplicate.

**Pull request merged**